File: dynamic programming/java/leetcode62_Unique_Paths.java
//A robot is located at the top-left corner of a m x n grid (marked 'Start' in the diagram below).
//
//The robot can only move either down or right at any point in time. The robot is trying to reach the bottom-right corner of the grid (marked 'Finish' in the diagram below).
//
//How many possible unique paths are there?

package wenxinjie;

public class leetcode62_Unique_Paths {
    public int uniquePaths(int m, int n) {
        int[] res = new int[m];
        res[0] = 1;
        for(int i = 0; i < n ; i ++ ){
            for (int j = 1 ; j < m ; j++){
                    res[j] += res[j-1];
            }
        }
        return res[m-1];    
    }
}

//Time: O(m*n)
//Space: O(n)
//Difficulty: medium

File: dynamic programming/python/leetcode62_Unique_Paths.py
# A robot is located at the top-left corner of a m x n grid (marked 'Start' in the diagram below).

# The robot can only move either down or right at any point in time. The robot is trying to reach the bottom-right corner of the grid (marked 'Finish' in the diagram below).

# How many possible unique paths are there?


class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        dp = [1 for _ in range(m)]
        for i in range(1,n):
            for j in range(1,m):
                dp[j] += dp[j-1]
        return dp[m-1]

# Time: O(m*n)
# Space: O(n)
# Difficulty: medium